content/en/blog/_posts/2024-09-25-swap-ga-graduation.md

```diff
@@ -0,0 +1,369 @@
+or NVMe.
+As swap might cause IO pressure, it is recommended to give a higher IO latency
+priority to system critical daemons. See the relevant section in the
+[best-practices](#best-practices) section below.
+
+### Memory-backed volumes
+
+On Linux nodes, memory-backed volumes (like [Secrets](/docs/concepts/configuration/secret/)
+and [emptyDirs](/docs/concepts/storage/volumes/#emptydir) with `medium: Memory`)
+are implemented with as `tmpfs` mount.
+The contents of such volumes should remain in memory at all times, hence should
+not be swapped to disk.
+To ensure the contents of such volumes remain in memory, the `noswap` tmpfs option
+is being used.
+
+The Linux kernel officially supports the `noswap` option from version 6.3 (more info
+can be found [here](/docs/reference/node/kernel-version-requirements/#requirements-other)).
+However, the different distributions often choose to backport this option to older
+Linux versions as well.
+
+In order to understand if the node supports the `noswap` option, Kubelet will do the following:
+* If the kernel's version is above 6.3 then the `noswap` option will be assumed to be supported.
+* Otherwise, kubelet would try to mount a dummy tmpfs with the `noswap` option at startup.
+  If kubelet fails with an error indicating of an unknown option, `noswap` will be assumed
+  to not be supported, hence will not be used.
+  If kubelet succeeds, the dummy tmpfs will be deleted and the `noswap` option will be used.
+  * If the `noswap` option is not supported, kubelet will emit a warning log entry,
+    then continue its execution.
+
+It is deeply encouraged to encrypt the swap space.
+See the [section above](#setting-up-encrypted-swap) with an example for setting unencrypted swap.
+However, handling encrypted swap is not within the scope of kubelet;
+rather, it is a general OS configuration concern and should be addressed at that level.
+It is the administrator's responsibility to provision encrypted swap to mitigate this risk.
+
+## Best practices
+
+### Disable swap for system-critical daemons
+
+During the testing phase and based on user feedback, it was observed that the performance
+of system-critical daemons and services might degrade.
+This implies that system daemons, including the kubelet, could operate slower than usual.
+If this issue is encountered, it is advisable to configure the cgroup of the system slice
+to prevent swapping (i.e., set `memory.swap.max=0`).
+
+### Protect system-critical daemons for io-latency
+
+Swap can increase the I/O load on a node.
+When memory pressure causes the kernel to rapidly swap pages in and out,
+system-critical daemons and services that rely on I/O operations may
+experience performance degradation.
+
+To mitigate this, it is recommended to prioritize the system slice in terms of I/O latency.
+This can be achieved by setting `io.latency` for the system slice,
+thereby granting it higher I/O priority.
```

Comment:
```suggestion
To mitigate this, it is recommended to prioritize the systemd slice named `system` in terms of I/O latency.
This can be achieved by setting `io.latency` for the system slice,
thereby granting it higher I/O priority.
```
Does this work, though? What is it being prioritized relative to?

Reply:
Yeah it does. As the docs say:
```
Specifies the number of milliseconds a process can wait before IO from other processes is given to it.
...
effectively prioritizing the job with the lowest io.latency setting.
```

See https://facebookmicrosites.github.io/cgroup2/docs/io-controller.html#protecting-workloads-with-iolatency for more info regarding this.

Reply:
Regarding the suggested change: I wanted to avoid mentioning systemd specifically because some environments might run without systemd, but still have a dedicated cgroup hierarchy for system processes.

WDYT?

Reply:
What other slice do you think is being deemphasized here? I worry that making that change won't actually improve things on a typical node.

Reply:
> What other slice do you think is being deemphasized here? 

The kubepods slice.

> I worry that making that change won't actually improve things on a typical node.

We tested this pretty extensively. I believe that in certain situations it is extremely helpful. For example, in situations where pods were swapping in-and-out extensively causing an IO pressure for system processes that are outside of kubelet's scope.

However if you think it's better to remove this section I won't insist. This is just a recommendation after all :)

Reply:
As far as I know, [in the Linux mainstream] only systemd has a slice concept. So either write for systemd users, or make it even more generic.